### PR TITLE
RBAC functionality

### PR DESCRIPTION
**Pull request by rajiteh**
Figured we'd merge our work together, if thats okay with you. We'll keep the discussion at https://github.com/kubernetes-incubator/kargo/pull/1351

* Add a flag "authorization_method", when set to "RBAC" enables role based access control.
* Add required cluster roles and bindings for kube-dns
* Patch tiller deployment to use a service account with proper credentials.
* Add a flag to regenerate kubernetes certs on the nodes.

### CODE REVIEW

File: docs/vars.md [hunk 1]
```diff
@@ -67,7 +67,11 @@ following default cluster paramters:
   OpenStack (default is unset)
 * *kube_hostpath_dynamic_provisioner* - Required for use of PetSets type in
   Kubernetes
-
+* *authorization_mode* - Set this to "RBAC" (upper-case, no quotes) 
+[to enable Role Based Access Control](https://kubernetes.io/docs/admin/authorization/rbac/)
+* *rotate_kubernetes_certs* - Set this to true to regenerate kubernetes Node certificates. *Warning: Will overwrite old certs.*
+
```
Comment (jwfang) [inline docs/vars.md]: maybe in a seperate PR ?

File: roles/kubernetes-apps/ansible/defaults/main.yml [hunk 2]
```diff
@@ -40,3 +40,11 @@ netchecker_server_memory_requests: 64M
...
 etcd_cert_dir: "/etc/ssl/etcd/ssl"
 canal_cert_dir: "/etc/canal/certs"
+
+# RBAC
+rbac_resources:
+  - clusterrole,
+  - clusterrolebinding,
+  - sa
+
+rbac_enabled: "{{ authorization_mode == 'RBAC' }}"
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: should we use *contains* to check for RBAC here, user may choose multiple authorizator.

File: roles/kubernetes-apps/ansible/tasks/main.yml [hunk 3]
```diff
@@ -13,12 +13,15 @@
...
+    - {name: kubedns, file: kubedns-sa.yml, type: sa}
     - {name: kubedns, file: kubedns-deploy.yml, type: deployment}
     - {name: kubedns, file: kubedns-svc.yml, type: svc}
+    - {name: kubedns-autoscaler, file: kubedns-autoscaler-sa.yml, type: sa}
+    - {name: kubedns-autoscaler, file: kubedns-autoscaler-clusterrole.yml, type: clusterrole}
+    - {name: kubedns-autoscaler, file: kubedns-autoscaler-clusterrolebinding.yml, type: clusterrolebinding}
     - {name: kubedns-autoscaler, file: kubedns-autoscaler.yml, type: deployment}
   register: manifests
-  when: dns_mode != 'none' and inventory_hostname == groups['kube-master'][0]
+  when: dns_mode != 'none' and inventory_hostname == groups['kube-master'][0] and (item.type not in rbac_resources or rbac_enabled)
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/tasks/main.yml]: nice catch, i didn't do this check.

File: roles/kubernetes-apps/ansible/templates/kubedns-deploy.yml [hunk 4]
```diff
@@ -114,4 +114,7 @@ spec:
         - containerPort: 8080
           protocol: TCP
       dnsPolicy: Default  # Don't use cluster DNS.
+{% if authorization_mode is defined and authorization_mode == "RBAC" %}
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/templates/kubedns-deploy.yml]: contains "RBAC"

File: roles/kubernetes-apps/ansible/templates/kubedns-deploy.yml [hunk 5]
```diff
@@ -114,4 +114,7 @@ spec:
...
+      serviceAccount: kube-dns
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/templates/kubedns-deploy.yml]: serviceAccount is deprecated, and only a alias/link to serviceAccountName.
see https://kubernetes.io/docs/resources-reference/v1.6/#podspec-v1-core

so maybe serviceAccount not needed ?

File: roles/kubernetes/master/defaults/main.yml [hunk 6]
```diff
@@ -64,4 +64,6 @@ apiserver_custom_flags: []
 
 controller_mgr_custom_flags: []
 
-scheduler_custom_flags: []
\ No newline at end of file
+scheduler_custom_flags: []
+
+authorization_mode: RBAC
```
Comment (jwfang) [inline roles/kubernetes/master/defaults/main.yml]: i think we should leave authorization no RBAC for compability reason, what do you think ?

File: roles/kubernetes-apps/helm/tasks/main.yml [hunk 7]
```diff
@@ -10,10 +10,25 @@
     mode: 0755
   register: helm_container
 
+- name: Helm | Configure tiller service account for RBAC
+  command: kubectl create serviceaccount tiller --namespace={{ system_namespace }}
```
Comment (jwfang) [inline roles/kubernetes-apps/helm/tasks/main.yml]: should we use the *kube* module and make all k8s resouces in files/templates ?

File: roles/kubernetes-apps/ansible/defaults/main.yml [hunk 8]
```diff
@@ -40,3 +40,11 @@ netchecker_server_memory_requests: 64M
 # SSL
 etcd_cert_dir: "/etc/ssl/etcd/ssl"
 canal_cert_dir: "/etc/canal/certs"
+
+# RBAC
+rbac_resources:
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: kubedns_rbac_resources ?

File: roles/kubernetes-apps/ansible/tasks/main.yml [hunk 9]
```diff
@@ -30,6 +33,7 @@
     filename: "{{kube_config_dir}}/{{item.item.file}}"
     state: "{{item.changed | ternary('latest','present') }}"
   with_items: "{{ manifests.results }}"
+  failed_when: manifests|failed and "Error from server (AlreadyExists)" not in manifests.msg
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/tasks/main.yml]: when did this (AlreadyExists) happen?

File: roles/kubernetes-apps/ansible/templates/kubedns-autoscaler-clusterrolebinding.yml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  name: cluster-proportional-autoscaler
+  namespace: {{ system_namespace }}
+subjects:
+  - kind: ServiceAccount
+    name: cluster-proportional-autoscaler
+    namespace: {{ system_namespace }}
+roleRef:
+  kind: ClusterRole
+  name: cluster-proportional-autoscaler
+  apiGroup: rbac.authorization.k8s.io
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/templates/kubedns-autoscaler-clusterrolebinding.yml]: no newline :)

File: roles/kubernetes/secrets/tasks/check-certs.yml [hunk 11]
```diff
@@ -10,8 +10,8 @@
 
 - name: "Check_certs | Set default value for 'sync_certs', 'gen_certs', and 'secret_changed'  to false"
   set_fact:
-    sync_certs: false
-    gen_certs: false
+    sync_certs: true
```
Comment (jwfang) [inline roles/kubernetes/secrets/tasks/check-certs.yml]: default true ?

File: docs/vars.md (hunk 1 above)
Comment (rajiteh) [inline docs/vars.md]: cert rotation is needed if we're upgrading an existing k8s cluster in to RBAC. (as the certs need to have specific common names and orgs. thats why i included it here. it could be a separate PR

File: roles/kubernetes-apps/ansible/tasks/main.yml (hunk 9 above)
Comment (rajiteh) [inline roles/kubernetes-apps/ansible/tasks/main.yml]: I started getting this error when I was re-running the `kubernetes-apps` role. It seems the `kube` module doesn't handle re-creation of ClusterRoles/Roles gracefully.

File: roles/kubernetes-apps/ansible/defaults/main.yml (hunk 2 above)
Comment (rajiteh) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: ^ Good idea. Will move.

File: roles/kubernetes-apps/ansible/defaults/main.yml (hunk 8 above)
Comment (rajiteh) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: the resource type list was not meant to be specific to kubedns, its universal to apiserver.

Comment (rajiteh) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: may be `apiserver_rbac_resources` ?

File: roles/kubernetes-apps/ansible/tasks/main.yml (hunk 9 above)
Comment (jwfang) [inline roles/kubernetes-apps/ansible/tasks/main.yml]: *kube* module seems to have *reloaded* state and *force* flag, but i didn't try them yet.

btw: i tried to implement a *patched* state for it, but it escape the json unless we use *unsafe_shell*. i guess run patch directly is giid enough now.

Comment (rajiteh) [inline roles/kubernetes-apps/ansible/tasks/main.yml]: I tried using the `force` flag and it didn't fix the issue. I'm not too familiar with custom ansible modules so there might be a trick to this somewhere.

`kubectl apply` should do the patching automatically on a file. im not sure if the module is using that approach. https://lukemarsden.github.io/docs/user-guide/kubectl/kubectl_apply/

**Comment by rajiteh** [comment]
@jwfang please take a look again. thanks!

File: roles/kubernetes/node/tasks/pre_upgrade.yml [hunk 12]
```diff
@@ -4,3 +4,7 @@
   args:
     creates: "/var/lib/cni"
   failed_when: false
+
+- name: "Pre-upgrade | Make sure to restart kubelet if certificates changed"
+  command: /bin/true
+  notify: restart kubelet if secrets changed
```
Comment (jwfang) [inline roles/kubernetes/node/tasks/pre_upgrade.yml]: new line :)

File: docs/vars.md [hunk 13]
```diff
@@ -67,7 +67,13 @@ following default cluster paramters:
   OpenStack (default is unset)
 * *kube_hostpath_dynamic_provisioner* - Required for use of PetSets type in
   Kubernetes
-
+* *authorization_mode* - A list of authorization modes that the apiserver should be configured. 
+Supported values are `['AlwaysAllow', 'RBAC']` (Default: `['AlwaysAllow']`)
```
Comment (jwfang) [inline docs/vars.md]: i guess we support all authorization modes, but only RBAC need special handing.

i am no native English speaker, but the rough idea here is that the user is free to choose from the AlwaysAllow/AlwaysDeny/ABRC/RBAC, and form the list from these four choice.

File: roles/kargo-defaults/defaults/main.yaml [hunk 14]
```diff
@@ -114,3 +114,23 @@ vault_deployment_type: docker
 k8s_image_pull_policy: IfNotPresent
 efk_enabled: false
 enable_network_policy: false
+
+## List of authorization plugins that must be configured for
```
Comment (jwfang) [inline roles/kargo-defaults/defaults/main.yaml]: same as above

File: roles/kubernetes-apps/ansible/defaults/main.yml [hunk 15]
```diff
@@ -40,3 +40,10 @@ netchecker_server_memory_requests: 64M
 # SSL
 etcd_cert_dir: "/etc/ssl/etcd/ssl"
 canal_cert_dir: "/etc/canal/certs"
+
+# RBAC specific resources that will be ignored when RBAC is not enabled.
+apiserver_rbac_resources:
```
Comment (jwfang) [inline roles/kubernetes-apps/ansible/defaults/main.yml]: i guess all of them are related to kubedns here ?

File: docs/vars.md (hunk 13 above)
Comment (rajiteh) [inline docs/vars.md]: ABAC, Webhook both needs extra config args to the apiserver (policy object, webhook urls, etc.) hence why I didn't want to confuse a user with allowing them to set them here and expect it to work. I feel like that work might be out of the scope for this PR.. I can change the description to indicate and remove the "supported values" part.

**Comment by rajiteh** [comment]
> we need to restart all components when its cert changed, not only kubelet, and do it in a fine grain and gentle way. and there maybe other corner cases we need to handle. that make it a seperate PR alone.

Good point. Something I hadn't realized. Removed 👍

File: docs/vars.md (hunk 13 above)
Comment (jwfang) [inline docs/vars.md]: sure, missed that, no ABRC now.

but can we make *authorization_mode* support both string and list, like *apiserver_custom_flags* ?

File: roles/kargo-defaults/defaults/main.yaml [hunk 16]
```diff
@@ -114,3 +114,20 @@ vault_deployment_type: docker
...
+## the k8s cluster.
+authorization_mode: ['AlwaysAllow']
+rbac_enabled: "{{ 'RBAC' in authorization_mode }}"
+
+
+ssl_ca_dirs: "[
```
Comment (jwfang) [inline roles/kargo-defaults/defaults/main.yaml]: should this also remove from this PR ?

File: docs/vars.md [hunk 17]
```diff
@@ -67,7 +67,8 @@ following default cluster paramters:
   OpenStack (default is unset)
 * *kube_hostpath_dynamic_provisioner* - Required for use of PetSets type in
   Kubernetes
-
+* *authorization_mode* - A list of authorization modes that the apiserver should be configured.
```
Comment (jwfang) [inline docs/vars.md]: maybe more doc here, and should we support both string and list like what *apiserver_custom_flags* does ?

File: roles/kubernetes/node/handlers/main.yml [hunk 18]
```diff
@@ -1,4 +1,9 @@
 ---
+- name: restart kubelet if secrets changed
```
Comment (jwfang) [inline roles/kubernetes/node/handlers/main.yml]: also not needed ?

File: roles/kubernetes/node/tasks/install.yml [hunk 19]
```diff
@@ -1,22 +1,8 @@
 ---
```
Comment (jwfang) [inline roles/kubernetes/node/tasks/install.yml]: this should be revert back, :)

File: roles/kubernetes/secrets/tasks/check-certs.yml [hunk 20]
```diff
@@ -25,7 +25,7 @@
 - name: "Check_certs | Set 'gen_certs' to true"
   set_fact:
     gen_certs: true
-  when: "not item in kubecert_master.files|map(attribute='path') | list"
+  when: "item not in (kubecert_master.files|map(attribute='path')|list)"
```
Comment (jwfang) [inline roles/kubernetes/secrets/tasks/check-certs.yml]: if not cert rotation, this changes not needed ?

Comment (jwfang) [inline roles/kubernetes/secrets/tasks/check-certs.yml]: if not cert rotation, this changes not needed ?

File: docs/vars.md (hunk 17 above)
Comment (rajiteh) [inline docs/vars.md]: > should we support both string and list like what apiserver_custom_flags does ?

I don't think thats needed because unlike `*custom_flags`,  its a single type (list of strings). Look at the how they've done`kube_apiserver_admission_control`.

File: roles/kubernetes/secrets/tasks/check-certs.yml (hunk 20 above)
Comment (rajiteh) [inline roles/kubernetes/secrets/tasks/check-certs.yml]: its just semantics. its more readable (imo) `<item> not in <list>` vs `not <item> in <list>`. that being said its not a part of the purpose of this PR
